File: Meziantou.ComicsReader.WebReader/src/services/apiClient.ts
import type {
  BooksResponse,
  BookResponse,
  PagesResponse,
  ReadingListResponse,
  ReadingListItemResponse,
  IndexingStatusResponse,
  UpdateReadingProgressRequest,
  VersionResponse,
} from '../types';

export const DEFAULT_REQUEST_TIMEOUT_MS = 15_000;
export const DEFAULT_IMAGE_REQUEST_TIMEOUT_MS = 60_000;

export class ApiError extends Error {
  status: number;

  constructor(status: number, message: string) {
    super(message);
    this.name = 'ApiError';
    this.status = status;
  }
}

// Thrown when the server cannot be reached (network failure or timeout)
export class ApiNetworkError extends Error {
  isTimeout: boolean;

  constructor(message: string, isTimeout: boolean, options?: ErrorOptions) {
    super(message, options);
    this.name = 'ApiNetworkError';
    this.isTimeout = isTimeout;
  }
}

export interface ApiClientOptions {
  requestTimeoutMs?: number;
  imageRequestTimeoutMs?: number;
}

function getHttpErrorMessage(response: Response): string {
  const status = `${response.status} ${response.statusText ?? ''}`.trim();
  switch (response.status) {
    case 401:
    case 403:
      return `Authentication failed (${status}). Check the access token in Settings.`;
    default:
      return `Server error (${status})`;
  }
}

export class ApiClient {
  private baseUrl: string;
  private token: string | null;
  private requestTimeoutMs: number;
  private imageRequestTimeoutMs: number;

  constructor(baseUrl: string, token: string | null = null, options: ApiClientOptions = {}) {
    this.baseUrl = baseUrl.replace(/\/+$/, '');
    this.token = token;
    this.requestTimeoutMs = options.requestTimeoutMs ?? DEFAULT_REQUEST_TIMEOUT_MS;
    this.imageRequestTimeoutMs = options.imageRequestTimeoutMs ?? DEFAULT_IMAGE_REQUEST_TIMEOUT_MS;
  }

  private getServerDisplayName(): string {
    return this.baseUrl || window.location.origin;
  }

  // Send a request and read its body, aborting if the whole operation exceeds the timeout
  private async send<T>(
    url: string,
    init: RequestInit,
    timeoutMs: number,
    readBody: (response: Response) => Promise<T>,
  ): Promise<T> {
    const controller = new AbortController();
    let timedOut = false;
    const timeoutId = setTimeout(() => {
      timedOut = true;
      controller.abort();
    }, timeoutMs);

    try {
      const response = await fetch(url, { ...init, signal: controller.signal });
      if (!response.ok) {
        throw new ApiError(response.status, getHttpErrorMessage(response));
      }

      return await readBody(response);
    } catch (error) {
      if (timedOut) {
        throw new ApiNetworkError(
          `The server ${this.getServerDisplayName()} did not respond within ${Math.round(timeoutMs / 1000)} seconds.`,
          true,
          { cause: error },
        );
      }

      // fetch rejects with a TypeError when the server is unreachable
      if (error instanceof TypeError) {
        throw new ApiNetworkError(`Unable to reach the server ${this.getServerDisplayName()}.`, false, { cause: error });
      }

      throw error;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  private async fetch<T>(path: string, options: RequestInit = {}): Promise<T> {
    const url = `${this.baseUrl}/api/v1${path}`;
    const headers = new Headers(options.headers);
    headers.set('Content-Type', 'application/json');

    if (this.token) {
      headers.set('Authorization', `Bearer ${this.token}`);
    }

    const text = await this.send(url, { ...options, headers }, this.requestTimeoutMs, response => response.text());

    // Handle empty responses
    if (!text) {
      return undefined as T;
    }

    return JSON.parse(text) as T;
  }

  private async fetchImage(url: string): Promise<Blob> {
    const headers = new Headers();

    if (this.token) {
      headers.set('Authorization', `Bearer ${this.token}`);
    }

    return this.send(url, { headers }, this.imageRequestTimeoutMs, response => response.blob());
  }

  // Books
  async getBooks(search?: string, filter?: string): Promise<BooksResponse> {
    const params = new URLSearchParams();
    if (search) params.set('search', search);
    if (filter) params.set('filter', filter);
    const queryString = params.toString();
    return this.fetch<BooksResponse>(`/books${queryString ? `?${queryString}` : ''}`);
  }

  async getBookInfo(path: string): Promise<BookResponse> {
    return this.fetch<BookResponse>(`/books/${encodeURIComponent(path)}/info`);
  }

  async getBookPages(path: string): Promise<PagesResponse> {
    return this.fetch<PagesResponse>(`/books/${encodeURIComponent(path)}/pages`);
  }

  getPageUrl(path: string, pageIndex: number): string {
    const url = `${this.baseUrl}/api/v1/books/${encodeURIComponent(path)}/pages/${pageIndex}`;
    if (this.token) {
      return `${url}?token=${encodeURIComponent(this.token)}`;
    }
    return url;
  }

  async getPage(path: string, pageIndex: number): Promise<Blob> {
    return this.fetchImage(`${this.baseUrl}/api/v1/books/${encodeURIComponent(path)}/pages/${pageIndex}`);
  }

  getCoverUrl(path: string): string {
    const url = `${this.baseUrl}/api/v1/books/${encodeURIComponent(path)}/cover`;
    if (this.token) {
      return `${url}?token=${encodeURIComponent(this.token)}`;
    }
    return url;
  }

  async getCover(path: string): Promise<Blob> {
    return this.fetchImage(`${this.baseUrl}/api/v1/books/${encodeURIComponent(path)}/cover`);
  }

  async markAsRead(path: string): Promise<void> {
    await this.fetch<void>(`/books/${encodeURIComponent(path)}/mark-as-read`, {
      method: 'POST',
    });
  }

  // Reading progress
  async getReadingList(includeCompleted: boolean = false): Promise<ReadingListResponse> {
    return this.fetch<ReadingListResponse>(`/reading-list?includeCompleted=${includeCompleted}`);
  }

  async getReadingListItem(path: string): Promise<ReadingListItemResponse | null> {
    try {
      return await this.fetch<ReadingListItemResponse>(`/reading-list/${encodeURIComponent(path)}`);
    } catch (error) {
      if (error instanceof ApiError && error.status === 404) {
        return null;
      }
      throw error;
    }
  }

  async updateReadingProgress(path: string, pageIndex: number): Promise<ReadingListResponse> {
    const request: UpdateReadingProgressRequest = { pageIndex };
    return this.fetch<ReadingListResponse>(`/reading-list/${encodeURIComponent(path)}`, {
      method: 'PUT',
      body: JSON.stringify(request),
    });
  }

  async removeFromReadingList(path: string): Promise<ReadingListResponse> {
    return this.fetch<ReadingListResponse>(`/reading-list/${encodeURIComponent(path)}`, {
      method: 'DELETE',
    });
  }

  // Indexing
  async getIndexingStatus(): Promise<IndexingStatusResponse> {
    return this.fetch<IndexingStatusResponse>('/indexing/status');
  }

  async triggerReindex(): Promise<void> {
    await this.fetch<void>('/indexing/reindex', {
      method: 'POST',
    });
  }

  // Version
  async getVersion(): Promise<VersionResponse> {
    return this.fetch<VersionResponse>('/version');
  }
}

// Singleton instance that gets updated when settings change
let apiClientInstance: ApiClient | null = null;

export function getApiClient(serverUrl: string, token: string | null): ApiClient {
  apiClientInstance = new ApiClient(serverUrl, token);
  return apiClientInstance;
}

export function getCurrentApiClient(): ApiClient | null {
  return apiClientInstance;
}

File: Meziantou.ComicsReader.iOS/ComicsReaderKit/Sources/ComicsReaderKit/APIClient.swift
import Foundation

public enum APIError: Error, LocalizedError, Equatable, Sendable {
    /// The server URL configured in the settings is not a valid http(s) URL
    case invalidServerURL(String)

    /// The server answered with a non-success status code
    case http(status: Int)

    /// The server cannot be reached (network failure or timeout)
    case network(server: String, isTimeout: Bool, timeoutSeconds: Int)

    /// The server answered with an unexpected payload
    case invalidResponse(String)

    public var errorDescription: String? {
        switch self {
        case .invalidServerURL(let url):
            "The server URL \"\(url)\" is invalid. Check the server URL in Settings."
        case .http(let status) where status == 401 || status == 403:
            "Authentication failed (\(Self.statusText(status))). Check the access token in Settings."
        case .http(let status):
            "Server error (\(Self.statusText(status)))"
        case .network(let server, true, let timeoutSeconds):
            "The server \(server) did not respond within \(timeoutSeconds) seconds."
        case .network(let server, false, _):
            "Unable to reach the server \(server)."
        case .invalidResponse(let message):
            "Invalid response from the server: \(message)"
        }
    }

    public var isNetworkError: Bool {
        if case .network = self {
            return true
        }

        return false
    }

    public var statusCode: Int? {
        if case .http(let status) = self {
            return status
        }

        return nil
    }

    private static func statusText(_ status: Int) -> String {
        "\(status) \(HTTPURLResponse.localizedString(forStatusCode: status).capitalized)"
    }
}

/// Operations exposed by the Comics Reader server
public protocol ComicsAPI: Sendable {
    func getBooks() async throws -> BooksResponse
    func getPage(path: String, pageIndex: Int) async throws -> Data
    func getCover(path: String) async throws -> Data
    func markAsRead(path: String) async throws
    func getReadingList(includeCompleted: Bool) async throws -> ReadingListResponse
    func getReadingListItem(path: String) async throws -> ReadingListItem?
    func updateReadingProgress(path: String, pageIndex: Int) async throws -> ReadingListResponse
    func removeFromReadingList(path: String) async throws -> ReadingListResponse
    func getIndexingStatus() async throws -> IndexingStatus
    func triggerReindex() async throws
    func getVersion() async throws -> VersionResponse
}

public final class APIClient: ComicsAPI {
    public static let defaultRequestTimeout: TimeInterval = 15
    public static let defaultImageRequestTimeout: TimeInterval = 60

    public let baseURL: URL
    private let token: String?
    private let session: URLSession
    private let requestTimeout: TimeInterval
    private let imageRequestTimeout: TimeInterval

    public init(
        serverURL: String,
        token: String?,
        session: URLSession = .shared,
        requestTimeout: TimeInterval = APIClient.defaultRequestTimeout,
        imageRequestTimeout: TimeInterval = APIClient.defaultImageRequestTimeout
    ) throws(APIError) {
        guard let baseURL = Self.normalizeServerURL(serverURL) else {
            throw .invalidServerURL(serverURL)
        }

        self.baseURL = baseURL
        self.token = token?.isEmpty == false ? token : nil
        self.session = session
        self.requestTimeout = requestTimeout
        self.imageRequestTimeout = imageRequestTimeout
    }

    /// Validates the server URL and removes trailing slashes
    public static func normalizeServerURL(_ value: String) -> URL? {
        var trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        while trimmed.hasSuffix("/") {
            trimmed.removeLast()
        }

        guard let url = URL(string: trimmed),
              let scheme = url.scheme?.lowercased(),
              scheme == "http" || scheme == "https",
              url.host() != nil else {
            return nil
        }

        return url
    }

    /// Encodes a value the same way as JavaScript's encodeURIComponent, so slashes in book paths are escaped
    public static func encodePathComponent(_ value: String) -> String {
        var allowed = CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")
        allowed.insert(charactersIn: "-_.!~*'()")
        return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
    }

    func makeURL(_ path: String) -> URL {
        // The path is already percent-encoded, so build the URL from the string to keep %2F as-is
        URL(string: baseURL.absoluteString + "/api/v1" + path)!
    }

    private func bookURL(_ bookPath: String, _ suffix: String) -> URL {
        makeURL("/books/\(Self.encodePathComponent(bookPath))\(suffix)")
    }

    private func readingListURL(_ bookPath: String) -> URL {
        makeURL("/reading-list/\(Self.encodePathComponent(bookPath))")
    }

    // Books

    public func getBooks() async throws -> BooksResponse {
        try await sendJSON(makeURL("/books"))
    }

    public func getBookPages(path: String) async throws -> PagesResponse {
        try await sendJSON(bookURL(path, "/pages"))
    }

    public func getPage(path: String, pageIndex: Int) async throws -> Data {
        try await send(bookURL(path, "/pages/\(pageIndex)"), timeout: imageRequestTimeout)
    }

    public func getCover(path: String) async throws -> Data {
        try await send(bookURL(path, "/cover"), timeout: imageRequestTimeout)
    }

    public func markAsRead(path: String) async throws {
        _ = try await send(bookURL(path, "/mark-as-read"), method: "POST", timeout: requestTimeout)
    }

    // Reading progress

    public func getReadingList(includeCompleted: Bool = false) async throws -> ReadingListResponse {
        try await sendJSON(makeURL("/reading-list?includeCompleted=\(includeCompleted)"))
    }

    public func getReadingListItem(path: String) async throws -> ReadingListItem? {
        do {
            return try await sendJSON(readingListURL(path))
        } catch APIError.http(status: 404) {
            return nil
        }
    }

    public func updateReadingProgress(path: String, pageIndex: Int) async throws -> ReadingListResponse {
        let body = try JSONCoding.makeEncoder().encode(UpdateReadingProgressRequest(pageIndex: pageIndex))
        return try await sendJSON(readingListURL(path), method: "PUT", body: body)
    }

    public func removeFromReadingList(path: String) async throws -> ReadingListResponse {
        try await sendJSON(readingListURL(path), method: "DELETE")
    }

    // Indexing

    public func getIndexingStatus() async throws -> IndexingStatus {
        try await sendJSON(makeURL("/indexing/status"))
    }

    public func triggerReindex() async throws {
        _ = try await send(makeURL("/indexing/reindex"), method: "POST", timeout: requestTimeout)
    }

    // Version

    public func getVersion() async throws -> VersionResponse {
        try await sendJSON(makeURL("/version"))
    }

    // Transport

    private func sendJSON<T: Decodable & Sendable>(_ url: URL, method: String = "GET", body: Data? = nil) async throws -> T {
        let data = try await send(url, method: method, body: body, timeout: requestTimeout)
        do {
            return try JSONCoding.makeDecoder().decode(T.self, from: data)
        } catch {
            throw APIError.invalidResponse(String(describing: error))
        }
    }

    private func send(_ url: URL, method: String = "GET", body: Data? = nil, timeout: TimeInterval) async throws -> Data {
        var request = URLRequest(url: url, timeoutInterval: timeout)
        request.httpMethod = method
        request.httpBody = body
        if body != nil {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }

        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let (data, response) = try await withTimeout(timeout) { [session, request] in
            try await session.data(for: request)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse("Not an HTTP response")
        }

        guard (200..<300).contains(httpResponse.statusCode) else {
            throw APIError.http(status: httpResponse.statusCode)
        }

        return data
    }

    // Abort the whole operation (not only when the connection is idle) if it exceeds the timeout
    private func withTimeout<T: Sendable>(_ timeout: TimeInterval, operation: @escaping @Sendable () async throws -> T) async throws -> T {
        let serverName = baseURL.absoluteString
        let timeoutError = APIError.network(server: serverName, isTimeout: true, timeoutSeconds: Int(timeout.rounded()))

        do {
            return try await withThrowingTaskGroup(of: T.self) { group in
                group.addTask {
                    try await operation()
                }
                group.addTask {
                    try await Task.sleep(for: .seconds(timeout))
                    throw timeoutError
                }

                defer { group.cancelAll() }
                return try await group.next()!
            }
        } catch let error as URLError {
            switch error.code {
            case .cancelled:
                throw CancellationError()
            case .timedOut:
                throw timeoutError
            default:
                throw APIError.network(server: serverName, isTimeout: false, timeoutSeconds: Int(timeout.rounded()))
            }
        }
    }
}
